### ISSUE TITLE
Simple formatter like Color.format("text") with BB Codes or Markdown Syntax?

### ISSUE THREAD

**Issue by YunaBraska** [dialex/JCDP#64]
**Describe the feature**
I wish to have a simple and modern formatter same as String.format to format variable text input.
Something simple, similar to BB Codes or Markdown?
With the current solution/structure, it's way too much effort.

**Expected behavior**
My wish would be something like:
* [BB Code style] Color.format("My custom text has [b]bold[/b] and [i]italic[/I] texts which you also can combine [b][I]combined[/I[/b]")
* [Markdown style] Color.format("My custom text has __bold__ and *italic* texts which you also can combine __*combined*__")

**Environment**
 - OS: MacOS/Linux

**Comment by dialex** [comment]
Hi Yuna,
That's an interesting idea 🤔  I don't think we can implement it neatly in Java though.

## 1. Shorthand

The simplest solution I can think of is creating a shorthand method like `b` for **Bold** that would return the same as running `colorize(text, BOLD())`. Sadly Java doesn't support concise variable string interpolation like JavaScript does:

```js
// JS
const text = `some ${b('bold text')} and some other text`
```

```java
// but in Java it would look like this
String text = String.format("some % and some other text", b("bold text"));
```

So [we would have to use `String.format`](https://stackoverflow.com/a/6431949/675577) which is a bit far off your expected usage. Because of that we would have the formatted parts of the original string all shifted to the end of the method. Check the example below:

```java
// Your expectation
String text = Color.format("My custom text has [b]bold[/b] and [i]italic[/I] texts which you also can [b][I]combinee[/I[/b]")

// Solution 1A: shorthand using String.format
String text = String.format("My custom text has %s and %s texts which you also can %s", b("bold"), i("italic")), b(i("combine")));

// Solution 1B: shorthand using string concatenation
String text = "My custom text has " + b("bold") + " and " + i("italic") + " texts which..."
```

## 2. Markdown parser

The other solution would be to implement a parser of Markdown to ANSI escape codes. It would take in a Markdown string, detect which substrings need to be formatted, and it would apply the necessary escape codes so that it displays formatted in the terminal.

```java
String text = mdFormat("My custom text has **bold** and _italic_ texts which you also can be _**combine**_");

String text = colorize(
  mdFormat("My custom text has **bold** and _italic_ texts which you also can be _**combine**_"),
  RED_TEXT()
);

String text = colorize(
  mdFormat("My custom text has **bold** and _italic_ texts which you also can be _**combine**_"),
  BOLD(), YELLOW_BACK()
);
```

That is not complex to make, but it requires time that I don't have right now. Also it needs thorough testing, to check how this new MD parser would behave together with JColor's existing `Attribute`s, like in the 2nd and specially the 3rd line in the example above.

## tl;dr

I understand the use case and think it's a neat idea. It requires some effort which I can put in the near future. If anyone wants to make a PR that implements the requirements above, I'll review it :)

**Comment by YunaBraska** [comment]
Fair enough 👍  Thx for looking into it. I also have not much time, which is why I have now a dirty solution like following but I would love to use your proper library ;)

Dirty Solution
```java
text
                .replace("[b]", "\033[1m")
                .replace("[/b]", ESC)
                .replace("[i]", "\033[3m")
                .replace("[/i]", ESC)
                .replace("[u]", "\033[4m")
                .replace("[/u]", ESC)
                .replace("[s]", "\033[9m")
                .replace("[/s]", ESC)
                ;
```

**Comment by dialex** [comment]
hehe yup, nice hack! It's not bulletproof but does the job
You can put all those `replace` inside a method and then you simply call it passing the text, like this

```java
public String mdFormat(String text) {
        return text
            .replace("[b]", "\033[1m")
            .replace("[/b]", ESC)
            .replace("[i]", "\033[3m")
            .replace("[/i]", ESC)
            .replace("[u]", "\033[4m")
            .replace("[/u]", ESC)
            .replace("[s]", "\033[9m")
            .replace("[/s]", ESC)
            ;
 }

// usage
String formatted = mdFormat("My custom text has [b]bold[/b] and [i]italic[/i]");
```
